testsuite/pytests/sli2py_mpi/mpi_test_wrapper.py

```diff
@@ -0,0 +1,197 @@
+# -*- coding: utf-8 -*-
+#
+# mpi_test_wrapper.py
+#
+# This file is part of NEST.
+#
+# Copyright (C) 2004 The NEST Initiative
+#
+# NEST is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 2 of the License, or
+# (at your option) any later version.
+#
+# NEST is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with NEST.  If not, see <http://www.gnu.org/licenses/>.
+
+import inspect
+import subprocess
+import tempfile
+import textwrap
+from pathlib import Path
+
+import pandas as pd
+import pytest
+from decorator import decorator
+
+
+class MPITestWrapper:
+    """
+    Base class that parses the test module to retrieve imports, test code and
+    test parametrization.
+    """
+
+    RUNNER = "runner.py"
+    SPIKE_LABEL = "spike-{}"
+    MULTI_LABEL = "multi-{}"
+    OTHER_LABEL = "other-{}"
+
+    RUNNER_TEMPLATE = textwrap.dedent(
+        """\
+        SPIKE_LABEL = '{spike_lbl}'
+        MULTI_LABEL = '{multi_lbl}'
+        OTHER_LABEL = '{other_lbl}'
+
+        {fcode}
+
+        if __name__ == '__main__':
+            {fname}({params})
+        """
+    )
+
+    def __init__(self, procs_lst, debug=False):
+        try:
+            iter(procs_lst)
+        except TypeError:
+            raise TypeError("procs_lst must be a list of numbers")
+
+        self._procs_lst = procs_lst
+        self._debug = debug
+        self._spike = None
+        self._multi = None
+        self._other = None
+
+    def _func_without_decorators(self, func):
+        return "".join(line for line in inspect.getsourcelines(func)[0] if not line.startswith("@"))
```

Comment:
Does this also work if the function definition is broken to multiple lines?
```
def foo(
    asdf,
    sdfg
):
```

Reply:
Does this actually work for functions that use fixtures?

Reply:
Good catch! It actually breaks if I split the decorator over two lines:
```python
@MPITestAssertEqual([1, 4],
                    debug=True)
def test_all_to_all():
```
because the `debug=True)` line then ends up in the runner. Need to think about that.

Reply:
Fixed now by basing the filtering on AST. I have also tried to improve documentation and moved material from README.md to module docstring for `mpi_test_wrapper`. And added a little example for parametrizing a test.